Fix TypeError: action.value[0] is undefined

dlebech opened the pull request:
This commit fixes an error that occurs when an ADD_RELATION action is
called with an empty value from a relationship field.

To be extra safe, the fix is applied both in the component and the
reducer.

- Add isEmpty check in SelectWrapper component when adding relations.
- Add array check in reducer function for ADD_RELATION.

Fix #9500

### What does it do?

- Add an extra check to the `action.value` item, thereby avoiding reading from an `undefined` object.

It might be the [`react-select` `onChange`](https://github.com/strapi/strapi/blob/cd809720ba002cb74e33b93b9b2abbc121aac6c9/packages/strapi-plugin-content-manager/admin/src/components/SelectMany/index.js#L85) event that sends empty values to the reducer when hitting the backspace key. Because the select box is an external dependency, changing the reducer was the easiest fix I could come up with.

### Why is it needed?

It fixes the issue described in #9500.

### How to test it?

Described in #9500

### Related issue(s)/PR(s)

#9500

strapi-cla commented:
[![CLA assistant check](https://cla.strapi.io/pull/badge/not_signed)](https://cla.strapi.io/strapi/strapi?pullRequest=9501) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.strapi.io/strapi/strapi?pullRequest=9501) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.strapi.io/check/strapi/strapi?pullRequest=9501) it.</sub>

soupette reviewed (changes requested):
Thank you for your contribution.

However, I think we should handle this case in the SelectWrapper component instead of the reducer.

Can you update [these lines](https://github.com/strapi/strapi/blob/1e9e02b4de065e48e8e43d7f78aa91446eaf0494/packages/strapi-plugin-content-manager/admin/src/components/SelectWrapper/index.js#L185-L187) instead?


```
const handleAddRelation = value => {
  if (!isEmpty(value)) {
    addRelation({ target: { name, value } });
  }
};
```

dlebech commented:
Thanks for looking at this @soupette .

Wouldn't you agree that it is incorrect of the reducer function to assume that `action.value` is an array [without checking for explicitly](https://github.com/strapi/strapi/blob/1e9e02b4de065e48e8e43d7f78aa91446eaf0494/packages/strapi-plugin-content-manager/admin/src/containers/EditViewDataManagerProvider/reducer.js#L66-L70)? So it seems to me that adding some kind of check there, like `Array.isArray` or checking for non-empty value would be a more safe strategy? After all, in the end it's the reducer's job to ensure the data is correct, isn't it?

Just my two cents. I'll take a look at your suggestion tonight.

soupette commented:
Well I agree, but it is also the component's job to send the correct data. In this particular case, do to the customisation the input should not send an `onChange` event when it is not needed, since we have overridden the behaviour of the react-select, we need to make sure it does not send improper events.

WDYT?

We could definitely add both logic by the way, but I really want to rely on the data sent by the components instead of handling errors that might come from a wrong implementation of the inputs.

dlebech commented:
@soupette Thanks for your swift response. Yeah that makes sense to me :+1: I'll look into adding both changes tonight then. Will take me a little while for me to test this, but I should have something ready for you by tomorrow. Hopefully that timeline works for you :slightly_smiling_face:

dlebech commented:
@soupette I've updated the PR with your requested changes to the `SelectWrapper` component.

I've also adjusted the reducer to explicitly check that the value being passed is a non-empty array. I considered using `isEmpty` here as well, but it seems it will return true for non-empty objects as well, so I though it would be good to be explicit about the array check. I saw that you are doing this in other places in the code base as well, but let me know if you want me to change it :slightly_smiling_face:

soupette reviewed (approved):
LGTM!

Thank you for your contribution!

dlebech commented:
Happy to help @soupette . I think something is off with codecov btw, I don't think I reduced coverage by 8% :sweat_smile: Do you want me to rebase the branch, or do you guys take it from here?
